common/djangoapps/track/contexts.py

```diff
@@ -12,26 +12,27 @@
 log = logging.getLogger(__name__)
 
 
-def course_context_from_url(url):
+def course_context_from_url(url, course_id_string=None):
     """
-    Extracts the course_context from the given `url` and passes it on to
-    `course_context_from_course_id()`.
+    If course_id_string string is not present, extracts it from the given `url`. Either way, then passes
+    it on to `course_context_from_course_id()`.
     """
     url = url or ''
-
-    match = COURSE_REGEX.match(url)
     course_id = None
-    if match:
-        course_id_string = match.group('course_id')
-        try:
-            course_id = CourseKey.from_string(course_id_string)
-        except InvalidKeyError:
-            log.warning(
-                'unable to parse course_id "{course_id}"'.format(
-                    course_id=course_id_string
-                ),
-                exc_info=True
-            )
+
+    if course_id_string is None:
+        match = COURSE_REGEX.match(url)
+        if match:
+            course_id_string = match.group('course_id')
+    try:
+        course_id = CourseKey.from_string(course_id_string)
+    except InvalidKeyError:
+        log.warning(
+            'unable to parse course_id "{course_id}"'.format(
```

Comment:
1. It seems like it might be useful to log the URL for later debugging? Is there anything else we could log that would help us narrow down where it's coming from?
2. Since the `course_id` parameter ultimately comes from user input, please escape it in some way (e.g. `repr` of it) and limit the length of what's put into the log. The URL is implicitly bounded nginx limits (8K?), but by accident or malice, someone can toss us a 10 MB `course_id`.

Reply:
attempting resolution in next push
1. logging url (truncated to...256 length) if course_id couldn't be parsed out (and was needed)
2. course_id wrapped in repr and truncated to 256